File: 백준/Silver/9012.괄호/괄호.java
import java.util.*;
import java.io.*;


public class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int n = Integer.parseInt(br.readLine());

        StringBuilder sb = new StringBuilder();
        Stack<Character> s = new Stack<>();
        for (int i = 0; i < n; i++) {
            boolean check = true;
            String str = br.readLine();
            s.clear();
            for(int j = 0; j <str.length();j++){
                char c = str.charAt(j);
                if(c == '('){
                    s.push(c);
                }
                else if(c == ')'){
                    if(s.isEmpty()){
                        check = false;
                        break;
                    }else{
                        s.pop();
                    }
                }
            }
            if(check && s.isEmpty()){
                sb.append("YES").append("\n");
            }else{
                sb.append("NO").append("\n");
            }
        }
        System.out.println(sb);
    }
}

File: 백준/Silver/9012. 괄호/괄호.py
import sys

T = int(sys.stdin.readline())


for _ in range(T):
    data = list(sys.stdin.readline().rstrip())

    stack = []
    for i in data:
        if i == '(':
            stack.append(i)
        elif i == ')':
            if len(stack) != 0 and stack[-1] == '(':
                stack.pop()
            else:
                stack.append(i)
    if len(stack) == 0:
        print("YES")
    else:
        print("NO")

